### PR TITLE
Implement Arrow conversions for char and token span arrays

### PR DESCRIPTION
**Pull request by BryanCutler**
Add Arrow conversions to/from char and token span arrays to enable saving DataFrames with these columns to feather files. This first implementation is not optimized for storage space and will save target text in the field metadata, possibly duplicated. Also, TokenSpanArrays will save a duplicate of the dependent CharSpanArrays in the same struct column. These could be improved in followup changes.

### CODE REVIEW

File: text_extensions_for_pandas/array/arrow_conversion.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+        pa.PyExtensionType.__init__(self, pa.struct(fields))
+
+    def __reduce__(self):
+        index_dtype = self.storage_type[self.BEGINS_NAME].type
+        metadata = self.storage_type[ArrowCharSpanType.BEGINS_NAME].metadata
+        return ArrowTokenSpanType, (index_dtype, metadata)
+
+
+def char_span_to_arrow(char_span):
```
Comment (frreiss) [inline text_extensions_for_pandas/array/arrow_conversion.py:102]: Missing type hints here. Should be `def char_span_to_arrow(char_span) -> pyarrow.ExtensionArray:`

File: text_extensions_for_pandas/array/arrow_conversion.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+    typ = ArrowCharSpanType(begins_array.type, char_span.target_text)
+    fields = list(typ.storage_type)
+
+    storage = pa.StructArray.from_arrays([begins_array, ends_array], fields=fields)
+
+    return pa.ExtensionArray.from_storage(typ, storage)
+
+
+def arrow_to_char_span(extension_array):
```
Comment (frreiss) [inline text_extensions_for_pandas/array/arrow_conversion.py:124]: Missing type hints.

File: text_extensions_for_pandas/array/arrow_conversion.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            raise ValueError("Only pyarrow.Array with a single chunk is supported")
+        extension_array = extension_array.chunk(0)
+
+    assert pa.types.is_struct(extension_array.storage.type)
+
+    # Get target text from the begins field metadata and decode string
+    metadata = extension_array.storage.type[ArrowCharSpanType.BEGINS_NAME].metadata
+    target_text = metadata[ArrowCharSpanType.TARGET_TEXT_KEY]
+    if isinstance(target_text, bytes):
+        target_text = target_text.decode()
```
Comment (frreiss) [inline text_extensions_for_pandas/array/arrow_conversion.py:143]: Should this line specify a specific encoding?

File: text_extensions_for_pandas/array/arrow_conversion.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    ends_array = extension_array.storage.field(ArrowCharSpanType.ENDS_NAME)
+
+    # Zero-copy convert arrays to numpy
+    begins = begins_array.to_numpy()
+    ends = ends_array.to_numpy()
+
+    return CharSpanArray(target_text, begins, ends)
+
+
+def token_span_to_arrow(token_span):
```
Comment (frreiss) [inline text_extensions_for_pandas/array/arrow_conversion.py:156]: Missing type hints.

File: text_extensions_for_pandas/array/arrow_conversion.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    extension array can be serialized and transferred with standard
+    Arrow protocols.
+
+    :param token_span: A TokenSpanArray to be converted
+    :return: pyarrow.ExtensionArray containing TokenSpan data
+    """
+    # Create array for begins, ends
+    token_begins_array = pa.array(token_span.begin_token)
+    token_ends_array = pa.array(token_span.end_token)
+    char_begins_array = pa.array(token_span.begin)
```
Comment (frreiss) [inline text_extensions_for_pandas/array/arrow_conversion.py:169]: This line of code will store the character-based begin offset of each *span* in `token_span`. You should be storing the character-based begin offset of each token in the underlying tokens behind `token_span`.

File: text_extensions_for_pandas/array/test_token_span.py [hunk 6]
```diff
@@ -292,5 +294,18 @@ def test_as_frame(self):
         self.assertEqual(len(df), len(arr))
 
 
+class TokenSpanArrayIOTests(ArrayTestBase):
+
+    def test_feather(self):
+        arr = self._make_spans_of_tokens()
```
Comment (frreiss) [inline text_extensions_for_pandas/array/test_token_span.py:300]: You need to create a second test case that creates `TokenSpan`s that cover more than one token per span. It would also be a good idea to test that zero-length spans are correctly serialized/deserialized.

File: text_extensions_for_pandas/array/arrow_conversion.py (hunk 5 above)
Comment (BryanCutler) [reply text_extensions_for_pandas/array/arrow_conversion.py:169]: Ahh ok, got it

File: text_extensions_for_pandas/array/arrow_conversion.py (hunk 3 above)
Comment (BryanCutler) [reply text_extensions_for_pandas/array/arrow_conversion.py:161]: I don't think it's necessary, pyarrow does the same I believe

**Comment by BryanCutler** [comment]
@frreiss I think this is working now, and hopefully I'm testing against all possible cases now. I had a crazy bug that took me a while to figure out. It was due to copy/paste a line with a comma at the end in the `ArrowTokenSpanType` class, which resulted in a SIGSEGV and no other stacktrace! Took me a while to figure out what was wrong  :sob:
